### ISSUE TITLE
Quorum calls invoked in-order might send messages to individual servers out-of-order

### ISSUE THREAD

**Issue by meling** [relab/gorums#16]
![qrpc](https://cloud.githubusercontent.com/assets/68231/25727096/a3416ba0-3127-11e7-8a2b-76b987ce87c2.png)

See the above sketch illustrating an important problem with synchronous quorum calls.

**The details:**
When a client performs what she thinks to be a synchronous quorum call by sending an RPC to all servers in a configuration, it is actually spawning a goroutine for each server. This is illustrated below for a `Read` quorum call.

```go
func (c *Configuration) Read(ctx context.Context, a *ReadRequest) (resp *State, err error) {
	...
	for _, n := range c.nodes {
		go callGRPCRead(ctx, n, a, replyChan)
	}
	...
```

```go
func callGRPCRead(ctx context.Context, node *Node, arg *ReadRequest, replyChan chan<- readReply) {
	err := grpc.Invoke(ctx, "/dev.Register/Read", arg, reply, node.conn)
	replyChan <- readReply{node.id, reply, err}
}
```

This `Read` quorum call can complete before all replies have been received, because only a subset of successful replies are needed to complete a quorum call. Exactly when enough replies have been received is determined by a quorum function, I won't explain the details here, except to say that in a typical majority quorum function, invoking a configuration with three servers, a majority of two replies is enough to complete a call. This has the potential to leave some goroutines running, or possibly even waiting to be started.

Thus, in rare cases it can happen that synchronous quorum call QC1 completes, but some of its goroutines haven't even started running yet. Moreover, since QC1 is complete, the client can perform the next quorum call QC2, and some of its goroutines may actually be scheduled before those belonging to QC1.

Note that it is okay for the goroutines of QC1 to linger after QC2 has started, as long as the RPC calls (TCP send call) belonging to QC1 are invoked before those belonging to QC2 for the same server.

Furthermore, often it is undesirable to cancel these lingering goroutines since it may cause some (slow) servers to become outdated with the rest. But there are some applications/protocols, where canceling these goroutines would be okay. However, that should not be the default since those that wish to cancel can do so through the context `ctx`.

**The problem it causes:**
A server can receive a message out-of-order, even though the underlying networking takes place over TCP. This is unexpected behavior.

**Is this relevant in practice?**
Yes. This is a problem in the Raft implementation as the server could see `m_102`, when expecting `m_101`, and the message is therefore discarded, because Raft expects to receive messages in order. Hence, there is no retransmission of `m_102` as the message was correctly received by the TCP layer, thus this server will fail to receive all succeeding messages as they are not the expected message `m_101`.

This has been experienced in a configuration running a cluster on a single machine, and possibly also in a LAN scenario when pressuring the leader, but that hasn't been confirmed yet. However, it happens only rarely, as it depends on the scheduling of goroutines, and so it is quite difficult to reproduce.

**Is it a problem only relevant to Raft?**
Raft requires strict message ordering per server, and does not tolerate holes in the log. This is not the case for protocols that can deal with multiple concurrent requests, such as MultiPaxos with α>1. That is algorithms that allows holes in the log. It is not clear if it is a problem in majority-based read-write storage systems, or other such systems, but it would probably require a combination of failures and this bug.

I expect that other protocols and systems may also not tolerate holes in their log, but can't think of any right now.

**Some possible solutions:**

1. Leave it as is, and document it. 
1. Implement a general fix for the problem that applies to all uses of quorum calls. This would be the prefer solution, but it may have ramifications:
   1. One possible solution is to limit concurrency across quorum calls (prevent pipelining).
   1. Another solution is to use a `WaitGroup` in a unorthodox way to prevent the problem with very high probability.
1. Another solution is to provide an option that those protocols that need strict ordering can use.
1. Instead of an option, we could define a separate quorum call type with strict ordering.

**Other concerns:**

We should have a reliable test case that can demonstrate and reproduce this bug. However, this seems to require interacting with the go scheduler.

**Comment by s111** [comment]
I guess, ideally, what we would want is a function like this:
```go
func InvokeStart(ctx context.Context, method string, args, reply interface{}, cc *ClientConn, opts ...CallOption) <-chan error
```
This would send the request on the wire and immediately return to the caller, the error can then be read on the error channel and the reply can be inspected. We could then build quorum calls that work correctly with TCP.

This is how `Invoke` is currently implemented:
```go
// Invoke sends the RPC request on the wire and returns after response is received.
// Invoke is called by generated code. Also users can call Invoke directly when it
// is really needed in their use cases.
func Invoke(ctx context.Context, method string, args, reply interface{}, cc *ClientConn, opts ...CallOption) error
```
That is, we are unable to implement our desired behavior without making changes to gRPC itself. As can be read from the function documentation, the request is sent on the wire and the function blocks until a response is received. Thus, any implementation that tries to sequence `grpc.Invoke` calls to individual servers (i.e., my channel based approach) will incur additional latency on every message sent, this is not good.

If we want TCP like behavior, I'm leaning towards an implementation using streams instead, as streams enforce the ordering properties that we seek.

**Comment by tormoder** [comment]
Initial test case added in e23441d.

**Comment by tormoder** [comment]
For now, all my attempts at adding explicit internal synchronization to ensure
ordering (without resorting to a non-concurrent version) has ended up being
equivalent to this:

Don't return/signal true from the quorum function before all responses have
been received, but with additional synchronization overhead.

Therefore, a question about Raft:

Why do we need to return from the QC after a quorum of replies have been
received, and why not wait for all replies before returning? I.e. what is
gained by returning before all replies have been received since we anyway want
to wait for slot x (QC-x) before starting y (QC-y). 

My suggestion is to use a Correctable if it is necessary to return before the
call as a whole is finished.  It would have two levels, `QUORUM` and `FINISHED`.
Then one could wait for QUORUM first and do an action (though I don't see what
that would be, therefore my question above) and finally wait for `FINISHED` (i.e.
`Done()`) before staring another call.

One possible solution is to document that:
* A regular QC that use cancel (implicitly) guarantees ordering.
* A regular QC that does not use cancel don't have any ordering guarantees.
* Applications requiring ordering semantics and need to return from the QC
  before every reply is received should use the technique described above.  

Side-note: This would by an argument for keeping the current basic
Correctable version, ref. #19.

**Comment by meling** [comment]
My preferred option is to try and see what can be achieved with a stream based approach, perhaps we can still mimick the traditional quorum call for the user. I haven't gotten to implementing it though, but it is certainly something I will try unless someone beats me to it...

**Comment by meling** [comment]
I'm playing with a structure like this (not done yet), where the `node` holds the client stream:
```go
	err := node.XReadClient.Send(arg)
	if err != nil {
		replyChan <- xReadReply{node.id, nil, err}
		return
	}

	for {
		reply, err := node.XReadClient.Recv()
		if err == io.EOF {
			return
		}
		replyChan <- xReadReply{node.id, reply, err}
		if err != nil {
			return
		}
	}
```

**Comment by meling** [comment]
Adding the PR for @s111 original strict ordering options implementation: https://github.com/relab/gorums-dev/pull/44

**Comment by meling** [comment]
Just adding reminder of test case that exposes the issue: `go test -tags order`. This should print something like this, if there is a problem (this is of course just an estimate since it can't be tested exhaustively):
```
--- FAIL: TestQuorumCallOrdering (10.02s)
	calltype_qc_ordering_test.go:75: server: message received out-of-order: got: 9, expected: 8
```
We don't run this as part of the full set of unit tests since we have not decided how to resolve this issue. I plan to look into this again soon, perhaps aiming to do a stream-based implementation.

**Comment by meling** [comment]
Restarted to think about this. I think the best solution is to establish a bidirectional stream connection. The `Node` type can host the stream, and each node also needs two channels, `sendQ` and `recvQ`. Messages to be sent on the stream is placed on the `sendQ` by our generated quorum call, that is, for example a `config.Read()` method, like this:

```go
func (conf *Configuration) Read(ctx, msg) Reply {
	for node := range conf.nodes {
		node.sendQ  <- msg
	}
	// Wait for replies on recvQ
}
```
(I assume `msg` is a pointer, so this loop will be fast. And there shouldn't be contention on the `sendQ` channel, since each node has its own, and each client typically will not be generating a bunch of concurrent quorum calls.)

Likewise, `config.Read()` also needs to select over the `recvQ`channel to get the replies, somewhat similar to the current implementation.

The `Node.connect()` method will establish connection to each node, and host goroutines for sending and receiving from the stream. Here is a very rough draft.

```go
	sendQ := make(chan string, 10)
	recvQ := make(chan *State, 10)
	stream, err := n.StorageClient.ReadStream(context.Background(), &ReadRequest{})
	if err != nil {
		return fmt.Errorf("stream creation failed: %v", err)
	}
	go func() {
		for {
			m := <-sendQ
			stream.SendMsg(m)
			reply, _ := stream.Recv()
			recvQ <- reply
			//TODO this goroutine must exit when connection closed
		}
	}()
```

A remaining concern is that we would need to generate server-side code as well, to do this:

```go
	for {
		msg, _ := stream.Recv()
		reply := srv.Read(msg)
		stream.SendMsg(reply)
	}
```

On the positive side of this need to generate server-side code, is that we could more easily support:

- Type safe metadata, which could be useful for passing `Configuration` and `QuorumSpec` objects between client and server
- Servers could also notify clients of events, such as configuration changes since they have persistent connections.

Not sure how easy or useful the latter will be.

Another concern is that we may be exposing the stream interface for our `Read()` method, which should ideally be hidden from the developer wanting to work with quorum calls. The only way I can think of to avoid this, is to generate our code from scratch, rather than relying on gRPC streams.

Other ideas are welcome.

**Comment by meling** [comment]
We should investigate if the [errgroup](https://godoc.org/golang.org/x/sync/errgroup) package can be used to solve the problem we have with out-of-order invocations.

**Comment by meling** [comment]
We now have an ordered implementation (thanks to @Raytar) based on an underlying stream. We should conduct a performance evaluation of this new implementation, and compare it with the current quorumcall implementation. If the performance impact is marginal, then there is really no need to keep the quorumcall implementation anymore, and the ordered one should take its place as the default. @Raytar can you guys take a stab at doing such a performance evaluation. It should involve remove nodes as well, e.g. in the BBChain cluster.

**Comment by Raytar** [comment]
Will do! What are the metrics we should look at? I am thinking of throughput vs latency, but are there other important ones?

**Comment by meling** [comment]
Yes, probably CPU load impact, number of goroutines, memory allocations, synchronization pressure and other micro benchmarks. I think pprof can provide some of this, but we may also consider to use the trace package (currently disabled for the ordered implementation). There are probably some other tools as well that can help...

**Comment by meling** [comment]
In relation to benchmarking this, it would be relevant to consider the impact of changing the type of `recvQ` to a `sync.Map`, as mentioned [here](https://github.com/relab/gorums/pull/55#discussion_r400499238).

**Comment by meling** [comment]
Closed by #75 (merged into master)
